IllegalArgumentException in SecurityTokenUtils (not enough width to fit value: 256/128)

App-Version 42202

Android-Version Android 6.0

Device Galaxy S7 Edge (hero2lte) 3


```
java.lang.RuntimeException: An error occurred while executing doInBackground()
	at android.os.AsyncTask$3.done(AsyncTask.java:309)
	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
	at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
	at java.util.concurrent.FutureTask.run(FutureTask.java:242)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
	at java.lang.Thread.run(Thread.java:818)
Caused by: java.lang.IllegalArgumentException: not enough width to fit value: 256/128
	at org.sufficientlysecure.keychain.securitytoken.SecurityTokenUtils.writeBits(SecurityTokenUtils.java:228)
	at org.sufficientlysecure.keychain.securitytoken.SecurityTokenUtils.createRSAPrivKeyTemplate(SecurityTokenUtils.java:87)
	at org.sufficientlysecure.keychain.securitytoken.SecurityTokenHelper.putKey(SecurityTokenHelper.java:513)
	at org.sufficientlysecure.keychain.securitytoken.SecurityTokenHelper.changeKey(SecurityTokenHelper.java:163)
	at org.sufficientlysecure.keychain.ui.SecurityTokenOperationActivity.doSecurityTokenInBackground(SecurityTokenOperationActivity.java:264)
	at org.sufficientlysecure.keychain.ui.base.BaseSecurityTokenActivity.handleSecurityToken(BaseSecurityTokenActivity.java:446)
	at org.sufficientlysecure.keychain.ui.base.BaseSecurityTokenActivity$1.doInBackground(BaseSecurityTokenActivity.java:172)
	at org.sufficientlysecure.keychain.ui.base.BaseSecurityTokenActivity$1.doInBackground(BaseSecurityTokenActivity.java:162)
	at android.os.AsyncTask$2.call(AsyncTask.java:295)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	... 4 more
```

--- a/OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenUtils.java
+++ b/OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenUtils.java
@@ -82,33 +82,35 @@ public class SecurityTokenUtils {
         template.write(new byte[]{(byte) 0x91, (byte) expLengthBytes});
         writeBits(data, secretKey.getPublicExponent(), expLengthBytes);
 
-        // Prime P, length 128
-        template.write(Hex.decode("928180"));
-        writeBits(data, secretKey.getPrimeP(), 128);
+        final int modLengthBytes = format.getModulusLength() / 8;
 
-        // Prime Q, length 128
+        // Prime P, length modLengthBytes / 2
+        template.write(Hex.decode("928180"));
+        writeBits(data, secretKey.getPrimeP(), modLengthBytes / 2);
+
+        // Prime Q, length modLengthBytes / 2
         template.write(Hex.decode("938180"));
-        writeBits(data, secretKey.getPrimeQ(), 128);
+        writeBits(data, secretKey.getPrimeQ(), modLengthBytes / 2);
 
 
         if (format.getAlgorithmFormat().isIncludeCrt()) {
-            // Coefficient (1/q mod p), length 128
+            // Coefficient (1/q mod p), length modLengthBytes / 2
             template.write(Hex.decode("948180"));
-            writeBits(data, secretKey.getCrtCoefficient(), 128);
+            writeBits(data, secretKey.getCrtCoefficient(), modLengthBytes / 2);
 
-            // Prime exponent P (d mod (p - 1)), length 128
+            // Prime exponent P (d mod (p - 1)), length modLengthBytes / 2
             template.write(Hex.decode("958180"));
-            writeBits(data, secretKey.getPrimeExponentP(), 128);
+            writeBits(data, secretKey.getPrimeExponentP(), modLengthBytes / 2);
 
-            // Prime exponent Q (d mod (1 - 1)), length 128
+            // Prime exponent Q (d mod (1 - 1)), length modLengthBytes / 2
             template.write(Hex.decode("968180"));
-            writeBits(data, secretKey.getPrimeExponentQ(), 128);
+            writeBits(data, secretKey.getPrimeExponentQ(), modLengthBytes / 2);
         }
 
         if (format.getAlgorithmFormat().isIncludeModulus()) {
-            // Modulus, length 256, last item in private key template
+            // Modulus, length modLengthBytes, last item in private key template
             template.write(Hex.decode("97820100"));
-            writeBits(data, secretKey.getModulus(), 256);
+            writeBits(data, secretKey.getModulus(), modLengthBytes);
         }
 
         // Bundle up

